Sprint 5

GoncharovNG opened the pull request:
На проверку

GoncharovNG committed:
Спринт номер 5 добавлены: протоколы, делегаты, алерта.

GoncharovNG committed:
sprint_5  сделано задание для отправки на ревью

helecta commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,26 @@
+    
+    func present(view controller: UIViewController, alert model: AlertModel) {
+        
+    let alert = UIAlertController(
```
Не забывай следить за отступами. Отступы помогают коду выглядеть аккуратнее, так читать код намного удобнее. Чтобы поправить оступы: можно выделить весь код `cmd+A` и зажать `contol+i`

GoncharovNG replied:
Исправлено!

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -30,151 +52,86 @@ final class MovieQuizViewController: UIViewController {
+                buttonText: result.buttonText,
+                completion: { [weak self] _ in
+                guard let self = self else { return }
+                self.currentQuestionIndex = 0
```
тут тоже отступов не хватает

GoncharovNG replied:
Исправлено

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -30,151 +52,86 @@ final class MovieQuizViewController: UIViewController {
             
-        // показываем всплывающее окно
-            self.present(alert, animated: true, completion: nil)
+            let alert = AlertPresenter()
```
лучше назвать не `alert`, а `alertPresenter`, потому что название `alert` воспринимается как ui элемент, а `alertPresenter` - объект, отвечающий за показ алерта.
И лучше объявить его один раз, а не создавать каждый раз новый. Сейчас пока это не сильно критично, но в будущем, если потребуется много алертов показывать и, например, сохранять состояния их, то его уже точно будет нужно объявлять один раз в `viewDidLoad`.

GoncharovNG replied:
Исправлено

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -30,151 +52,86 @@ final class MovieQuizViewController: UIViewController {
-        let image: String
-        let text: String
-        let correctAnswer: Bool
+    override func viewDidLoad() {
```
Стоит организовывать код следующим образом:
- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -30,151 +52,86 @@ final class MovieQuizViewController: UIViewController {
+        }
+    }
+}
 /*
```
Закомментированный код стоит удалить. Это делает код более аккуратным и легко читаемым.

GoncharovNG replied:
Сделано)

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -30,151 +52,86 @@ final class MovieQuizViewController: UIViewController {
+            statisticService.store(correct: correctAnswers, total: questionsAmount)
+            statisticService.gamesCount += 1
+            let text = correctAnswers == questionsAmount ?
+                    "Поздравляем, Вы ответили на 10 из 10!" : """
```
В макете другой формат текста, правильнее сделать как в макете. (Ваш результат ...)

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -30,151 +52,86 @@ final class MovieQuizViewController: UIViewController {
+                    "Поздравляем, Вы ответили на 10 из 10!" : """
+                    "Вы ответили на \(correctAnswers) из 10
+                    Количество сыгранных квизов: \(statisticService.gamesCount)
+                    Рекорд: \(statisticService.bestGame.correct)/\(statisticService.bestGame.total) (\(statisticService.bestGame.date))
```
Формат даты не такой как в макете, это нужно исправить.
Обрати внимание, у тебя в Helpers есть Date+Extensions, он поможет тебе с верным форматом даты.
Можешь почитать про DateFormatter [тут](https://developer.apple.com/documentation/foundation/dateformatter)

helecta commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,66 @@
+        let question = questions[safe: index]
+        delegate?.didReceiveNextQuestion(question: question)
+    }
+        init(delegate: QuestionFactoryDelegate?) {
```
Стоит организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+    let date: Date
+    
+    func reviewRecord(currentCorrect: Int, recordCorrect: Int ) -> Bool {
+     currentCorrect > recordCorrect
```
отступы

helecta replied:
1. Для методов, которые возвращают bool, в названии лучше использовать в начале is `isBigger(currentCorrect: Int, newCorrect: Int) -> Bool`. Тогда другим разработчикам, когда они будут читать твой код, сразу будет понятно какой тип возвращает метод. Подробнее [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming).
2. Т.к это метод класса GameRecord, то особо нет смысле передавать currentCorrect, лучше использовать self.correct:
`self.correct > recordCorrect`
Тогда и назвать можно было бы так:
`isCorrectBigger(newCorrect: Int) -> Bool`.
И при вызове код будет выглядеть так: `record.isCorrectBigger(newCorrect: newRecord.correct)`
3. Но вообще, тут удобнее использовать протокол Comparable. Как почитать можно [в этой статье](https://www.hackingwithswift.com/example-code/language/how-to-conform-to-the-comparable-protocol).

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,26 +1,48 @@
+    private let questionsAmount: Int = 10
+    private var questionFactory: QuestionFactoryProtocol?
+    private var currentQuestion: QuizQuestion?
+    private var statisticService = StatisticServiceImplementation()
```
По тз переменная `statisticService` должна быть типа `StatisticService`

GoncharovNG replied:
Исправлено

helecta replied:
Ты переименовал `statisticService` в `StatisticService`, а имелось в виду что надо сделать вот что:
`private var statisticService: StatisticService = StatisticServiceImplementation()`
я приму изменения, только ты это обязательно исправь, а то с большой буквы переменный не принято называть

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+        
+        let correctAnswers = (userDefaults.value(forKey: Keys.correct.rawValue) as? Int) ?? 0
+        let newCorrectAnswers = correctAnswers + count
+        userDefaults.set(newCorrectAnswers, forKey: Keys.correct.rawValue)
```
Давай вынесем в отдельные свойства `correct` и `total`.
Для них реализуем get и set.
Т.е:
```
var total: Double {
        get {
            return userDefaults.double(forKey: Keys.total.rawValue)
        }
        set {
            userDefaults.set(newValue, forKey: Keys.total.rawValue)
        }
    }
```

Тогда сможем аккуратно реализовать работу с `UserDefaults`. В `store` не будет прямого обращения к  `UserDefaults`

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    var gamesCount: Int {
+        get {
+            guard let data = userDefaults.data(forKey: Keys.gamesCount.rawValue), let gamesCount = try?
```
Обрати внимание UserDefaults поддерживает работу с примитивными типами [для чтения](https://developer.apple.com/documentation/foundation/userdefaults/1407405-integer) и [записи](https://developer.apple.com/documentation/foundation/userdefaults/1413614-set).

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+           let record = try? JSONDecoder().decode(GameRecord.self, from: bestGame) {
+            if record.reviewRecord(currentCorrect: count, recordCorrect: record.correct) {
+                let newRecord = GameRecord(correct: count, total: amount, date: Date())
+                let data = try? JSONEncoder().encode(newRecord)
```
У тебя реализован `get` и `set` для `bestGame`, поэтому тут можно просто использовать `self.bestGame`, когда нужно работать с сохраненной лучшей игрой.
`if self.bestGame.reviewRecord(currentCorrect: count, recordCorrect: self.bestGame.correct) {`

На самом деле слово self, тут необязательно использовать(но только обязательно удали строчку 
`if let bestGame = userDefaults.data(forKey: Keys.bestGame.rawValue),`
тогда можно будет просто 
`if bestGame.reviewRecord(currentCorrect: count, recordCorrect: bestGame.correct) {`

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+                let data = try? JSONEncoder().encode(newRecord)
+                userDefaults.set(data, forKey: Keys.bestGame.rawValue)
+            }
+        } else {
```
Это все тоже можно удалить, т.к первый if не нужен, если будешь пользоваться свойством `self.bestGame`

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,241 +1,134 @@
+    private let questionsAmount: Int = 10
+    private var questionFactory: QuestionFactoryProtocol?
+    private var currentQuestion: QuizQuestion?
+    private var StatisticService = StatisticServiceImplementation()
```
Ты переименовал `statisticService` в `StatisticService`, а имелось в виду что надо сделать вот что:
`private var statisticService: StatisticService = StatisticServiceImplementation()`
я приму изменения, только ты это обязательно исправь, а то с большой буквы переменный не принято называть

GoncharovNG committed:
Исправлены критические замечания для принятия проекта
Рекомендации, не влияющие на результат проверки исправлены не все
После прилета займусь ими (спасибо за рекомендации ))
